File: packages/core/android/src/newarch/kotlin/com/datadog/reactnative/DdLogs.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.reactnative

import com.facebook.fbreact.specs.NativeDdLogsSpec
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap

/**
 * The entry point to use Datadog's Logs feature.
 */
class DdLogs(
    reactContext: ReactApplicationContext,
    datadogWrapper: DatadogWrapper = DatadogSDKWrapper()
) : NativeDdLogsSpec(reactContext) {

    private val implementation = DdLogsImplementation(datadog = datadogWrapper)

    override fun getName(): String = DdLogsImplementation.NAME

    /**
     * Send a log with Debug level.
     * @param message The message to send.
     * @param context The additional context to send.
     */
    @ReactMethod
    override fun debug(message: String, context: ReadableMap, promise: Promise) {
        implementation.debug(message, context, promise)
    }

    /**
     * Send a log with Info level.
     * @param message The message to send.
     * @param context The additional context to send.
     */
    @ReactMethod
    override fun info(message: String, context: ReadableMap, promise: Promise) {
        implementation.info(message, context, promise)
    }

    /**
     * Send a log with Warn level.
     * @param message The message to send.
     * @param context The additional context to send.
     */
    @ReactMethod
    override fun warn(message: String, context: ReadableMap, promise: Promise) {
        implementation.warn(message, context, promise)
    }

    /**
     * Send a log with Error level.
     * @param message The message to send.
     * @param context The additional context to send.
     */
    @ReactMethod
    override fun error(message: String, context: ReadableMap, promise: Promise) {
        implementation.error(message, context, promise)
    }

    /**
     * Send a log containing an error with Debug level.
     * @param message The message to send.
     * @param errorKind The error kind to send.
     * @param errorMessage The error message to send.
     * @param stacktrace The error stacktrace to send.
     * @param context The additional context to send.
     */
    @Suppress("LongParameterList")
    @ReactMethod
    override fun debugWithError(
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stacktrace: String?,
        context: ReadableMap,
        promise: Promise
    ) {
        implementation.debugWithError(message, errorKind, errorMessage, stacktrace, context, promise)
    }

    /**
     * Send a log with containing an error with Info level.
     * @param message The message to send.
     * @param errorKind The error kind to send.
     * @param errorMessage The error message to send.
     * @param stacktrace The error stacktrace to send.
     * @param context The additional context to send.
     */
    @Suppress("LongParameterList")
    @ReactMethod
    override fun infoWithError(
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stacktrace: String?,
        context: ReadableMap,
        promise: Promise
    ) {
        implementation.infoWithError(message, errorKind, errorMessage, stacktrace, context, promise)
    }

    /**
     * Send a log containing an error with Warn level.
     * @param message The message to send.
     * @param errorKind The error kind to send.
     * @param errorMessage The error message to send.
     * @param stacktrace The error stacktrace to send.
     * @param context The additional context to send.
     */
    @Suppress("LongParameterList")
    @ReactMethod
    override fun warnWithError(
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stacktrace: String?,
        context: ReadableMap,
        promise: Promise
    ) {
        implementation.warnWithError(message, errorKind, errorMessage, stacktrace, context, promise)
    }

    /**
     * Send a log containing an error with Error level.
     * @param message The message to send.
     * @param errorKind The error kind to send.
     * @param errorMessage The error message to send.
     * @param stacktrace The error stacktrace to send.
     * @param context The additional context to send.
     */
    @Suppress("LongParameterList")
    @ReactMethod
    override fun errorWithError(
        message: String,
        errorKind: String?,
        errorMessage: String?,
        stacktrace: String?,
        context: ReadableMap,
        promise: Promise
    ) {
        implementation.errorWithError(message, errorKind, errorMessage, stacktrace, context, promise)
    }
}

File: packages/react-native-vega/kepler/turbo-modules/DdLogs.cpp
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
//
// This product includes software developed at Datadog
// (https://www.datadoghq.com/). Copyright 2016-Present Datadog, Inc.

#include "DdLogs.h"

#include <thread>
#include <utility>

#include "AttributeConverter.h"
#include "DatadogGlobalState.h"
#include "datadog/logging.hpp"

using namespace com::amazon::kepler::turbomodule;

namespace DdLogsTurboModule {
namespace {

enum class Level { Debug, Info, Warn, Error };

Promise emitLog(
    Level level,
    std::string message,
    JSObject context,
    std::string errorKind = "",
    std::string errorMessage = "",
    std::string stacktrace = ""
) {
    return Promise([
        level,
        message = std::move(message),
        context = std::move(context),
        errorKind = std::move(errorKind),
        errorMessage = std::move(errorMessage),
        stacktrace = std::move(stacktrace)
    ](const std::shared_ptr<Promise>& promise) {
        std::thread([
            promise,
            level,
            message,
            context,
            errorKind,
            errorMessage,
            stacktrace
        ]() {
            auto logger =
                datadog_rn_vega::DatadogGlobalState::getInstance().getLogger();
            if (logger) {
                const auto attributes =
                    datadog_rn_vega::jsObjectToAttribute(context);
                const datadog::LogError logError{
                    errorMessage, errorKind, stacktrace};

                switch (level) {
                    case Level::Debug:
                        logger->Debug(message, logError, attributes);
                        break;
                    case Level::Info:
                        logger->Info(message, logError, attributes);
                        break;
                    case Level::Warn:
                        logger->Warn(message, logError, attributes);
                        break;
                    case Level::Error:
                        logger->Error(message, logError, attributes);
                        break;
                }
            }
            promise->resolve(true);
        }).detach();
    });
}

}  // namespace

DdLogs::DdLogs() {}
DdLogs::~DdLogs() noexcept {}

Promise DdLogs::debug(std::string message, JSObject context) {
    return emitLog(Level::Debug, std::move(message), std::move(context));
}

Promise DdLogs::info(std::string message, JSObject context) {
    return emitLog(Level::Info, std::move(message), std::move(context));
}

Promise DdLogs::warn(std::string message, JSObject context) {
    return emitLog(Level::Warn, std::move(message), std::move(context));
}

Promise DdLogs::error(std::string message, JSObject context) {
    return emitLog(Level::Error, std::move(message), std::move(context));
}

Promise DdLogs::debugWithError(
    std::string message,
    std::string errorKind,
    std::string errorMessage,
    std::string stacktrace,
    JSObject context
) {
    return emitLog(
        Level::Debug,
        std::move(message),
        std::move(context),
        std::move(errorKind),
        std::move(errorMessage),
        std::move(stacktrace));
}

Promise DdLogs::infoWithError(
    std::string message,
    std::string errorKind,
    std::string errorMessage,
    std::string stacktrace,
    JSObject context
) {
    return emitLog(
        Level::Info,
        std::move(message),
        std::move(context),
        std::move(errorKind),
        std::move(errorMessage),
        std::move(stacktrace));
}

Promise DdLogs::warnWithError(
    std::string message,
    std::string errorKind,
    std::string errorMessage,
    std::string stacktrace,
    JSObject context
) {
    return emitLog(
        Level::Warn,
        std::move(message),
        std::move(context),
        std::move(errorKind),
        std::move(errorMessage),
        std::move(stacktrace));
}

Promise DdLogs::errorWithError(
    std::string message,
    std::string errorKind,
    std::string errorMessage,
    std::string stacktrace,
    JSObject context
) {
    return emitLog(
        Level::Error,
        std::move(message),
        std::move(context),
        std::move(errorKind),
        std::move(errorMessage),
        std::move(stacktrace));
}

}  // namespace DdLogsTurboModule
